File: 211/211.go
package beat_the_leet

type WordDictionary struct {
	end      bool
	children map[rune]*WordDictionary
}

func Constructor() WordDictionary {
	return WordDictionary{
		children: make(map[rune]*WordDictionary),
	}
}

func (wd *WordDictionary) AddWord(word string) {
	child := wd
	for _, r := range word {
		c, ok := child.children[r]
		if !ok {
			wd := Constructor()
			child.children[r] = &wd
			c = child.children[r]
		}

		child = c
	}
	child.end = true
}

func (wd *WordDictionary) Search(word string) bool {
	var ok bool
	child := wd
	for n, r := range word {
		if r == '.' {
			for _, c := range child.children {
				if found := c.Search(word[n+1:]); found {
					return true
				}
			}
			return false
		} else {
			child, ok = child.children[r]
			if !ok {
				return false
			}
		}
	}
	return child.end
}

File: 211/211.js
function Node(value) {
  this.value = value;
  this.endOfWord = false;
  this.children = {};
}

class WordDictionary {
  constructor() {
    this.root = new Node(null);
  }

  addWord(word) {
    let current = this.root;
    for (let i = 0; i < word.length; i += 1) {
      const letter = word[i];
      if (!current.children[letter]) {
        current.children[letter] = new Node(letter);
      }
      current = current.children[letter];
    }
    current.endOfWord = true;
  }

  search(word, root = undefined) {
    let current = root ?? this.root;
    for (let i = 0; i < word.length; i += 1) {
      const letter = word[i];
      if (letter === ".") {
        const childKeys = Object.keys(current.children);
        for (let j = 0; j < childKeys.length; j += 1) {
          if (this.search(word.slice(i + 1), current.children[childKeys[j]])) {
            return true;
          }
        }
        return false;
      } else {
        if (!current.children[letter]) {
          return false;
        }
        current = current.children[letter];
      }
    }
    return current.endOfWord;
  }
};
